### PR TITLE
Edit find to enable finding people by course code

### PR DESCRIPTION
**Pull request by Hibeom0929**
Pass all the words from the input to both NameContainsKeywordPredict and CourseContainsKeywordPredict.
Possible improvement:
- if Name reads a word as name, Course will not read that word again.
- For now, the person is detected when the full word matches any word in the name or course. ex) find Alic will not list Alice, find cs2103 will not list cs2103T. I think it is fine for the name but for the course code, not sure if approving short cut is better or not.

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 1]
```diff
@@ -25,9 +26,10 @@ public FindCommand parse(String args) throws ParseException {
                     String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
         }
 
-        String[] nameKeywords = trimmedArgs.split("\\s+");
+        String[] Keywords = trimmedArgs.split("\\s+");
```
Comment (E0735389) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:29]: This violates the naming convention

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 2]
```diff
@@ -48,13 +56,14 @@ public boolean equals(Object other) {
...
-        return predicate.equals(otherFindCommand.predicate);
+        return nPredicate.equals(otherFindCommand.nPredicate) && cPredicate.equals(otherFindCommand.cPredicate);
     }
 
     @Override
     public String toString() {
         return new ToStringBuilder(this)
-                .add("predicate", predicate)
+                .add("predicate", nPredicate)
+                .add("predicate", cPredicate)
```
Comment (E0735389) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:66]: This `toString` representation is very undescriptive.

**Review by E0735389 — CHANGES_REQUESTED** [review]
Apart from these minor changes pointed out, overall there appears to be some unnecessary duplication of logic in `NameContainsKeywordPredicate` and `CourseContainsKeywordPredicate`. How does the rest think about the design?

**Comment by Hibeom0929** [comment]
> Apart from these minor changes pointed out, overall there appears to be some unnecessary duplication of logic in `NameContainsKeywordPredicate` and `CourseContainsKeywordPredicate`. How does the rest think about the design?

Tbh, they are basically the same. Only the difference is one is Name and one is Course.

File: src/main/java/seedu/address/model/person/CourseContainsKeywordsPredicate.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    private final List<String> keywords;
+
+    public CourseContainsKeywordsPredicate(List<String> keywords) {
+        this.keywords = keywords;
+    }
+
+    @Override
+    public boolean test(Person person) {
+        return keywords.stream()
+                .anyMatch(keyword -> StringUtil.containsWordIgnoreCase(person.getCourse().value, keyword));
```
Comment (E0735389) [inline src/main/java/seedu/address/model/person/CourseContainsKeywordsPredicate.java:22]: Thinking about it, this is needlessly redundant as well. Isn't a course code always a single word anyway?

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 4]
```diff
@@ -48,13 +56,14 @@ public boolean equals(Object other) {
...
-        return predicate.equals(otherFindCommand.predicate);
+        return nPredicate.equals(otherFindCommand.nPredicate) && cPredicate.equals(otherFindCommand.cPredicate);
     }
 
     @Override
     public String toString() {
         return new ToStringBuilder(this)
-                .add("predicate", predicate)
+                .add("name predicate", nPredicate)
+                .add("course predicate", cPredicate)
```
Comment (E0735389) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:66]: The point is that the fieldName should be **the name of the field**, according to `ToStringBuilder`'s documentation.

Also that the field name appears to be needlessly obscure as well, but not a big deal.

**Review by E0735389 — COMMENTED** [review]
>  they are basically the same. Only the difference is one is Name and one is Course.

That's the point (nevertheless I don't really see a very good way to avoid the duplication of code either). We see what people may say in the forum `https://github.com/nus-cs2103-AY2324S2/forum/issues/619`

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 5]
```diff
@@ -25,9 +26,10 @@ public FindCommand parse(String args) throws ParseException {
                     String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
         }
 
-        String[] nameKeywords = trimmedArgs.split("\\s+");
+        String[] keywords = trimmedArgs.split("\\s+");
 
-        return new FindCommand(new NameContainsKeywordsPredicate(Arrays.asList(nameKeywords)));
+        return new FindCommand(new NameContainsKeywordsPredicate(Arrays.asList(keywords)),
+                               new CourseContainsKeywordsPredicate(Arrays.asList((keywords))));
```
Comment (E0735389) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:32]: Redundant parentheses here.

**Review by E0735389 — CHANGES_REQUESTED** [review]

**Comment by Hibeom0929** [comment]
> > they are basically the same. Only the difference is one is Name and one is Course.
> 
> That's the point (nevertheless I don't really see a very good way to avoid the duplication of code either). We see what people may say in the forum `https://github.com/nus-cs2103-AY2324S2/forum/issues/619`

Based on the reply in the forum, I think we can have both NameContainsKeywordPredicate and CourseContainsKeywordPredicate because it is clear and easy to be tested.

**Review by E0735389 — APPROVED** [review]

**Pull request merged**